langchain/langchain_interceptor.py

```diff
@@ -0,0 +1,174 @@
+        self, next: temporalio.worker.ActivityInboundInterceptor
+    ) -> temporalio.worker.ActivityInboundInterceptor:
+        return _LangChainContextPropagationActivityInboundInterceptor(next)
+
+    def workflow_interceptor_class(
+        self, input: temporalio.worker.WorkflowInterceptorClassInput
+    ) -> Type[_LangChainContextPropagationWorkflowInboundInterceptor]:
+        return _LangChainContextPropagationWorkflowInboundInterceptor
+
+class _LangChainContextPropagationClientOutboundInterceptor(
+    temporalio.client.OutboundInterceptor
+):
+    def __init__(
+        self,
+        next: temporalio.client.OutboundInterceptor,
+        payload_converter: temporalio.converter.PayloadConverter,
+    ) -> None:
+        super().__init__(next)
+        self._payload_converter = payload_converter
+
+    async def start_workflow(
+        self, input: temporalio.client.StartWorkflowInput
+    ) -> temporalio.client.WorkflowHandle[Any, Any]:
+        with trace(name=f"start_workflow:{input.workflow}"):
+            set_header_from_context(input, self._payload_converter)
+            return await super().start_workflow(input)
+
+class _LangChainContextPropagationActivityInboundInterceptor(
+    temporalio.worker.ActivityInboundInterceptor
+):
+    async def execute_activity(
+        self, input: temporalio.worker.ExecuteActivityInput
+    ) -> Any:
+        if isinstance(input.fn, str):
+            name = input.fn
+        elif callable(input.fn):
+            defn = temporalio.activity._Definition.from_callable(input.fn)
+            if not defn or not defn.name:
+                name = "unknown"
+            name = defn.name
+
+        with context_from_header(input, temporalio.activity.payload_converter()):
+            with trace(name=f"execute_activity:{name}"):
+                return await self.next.execute_activity(input)
+
+class _LangChainContextPropagationWorkflowInboundInterceptor(
+    temporalio.worker.WorkflowInboundInterceptor
+):
+    def init(self, outbound: temporalio.worker.WorkflowOutboundInterceptor) -> None:
+        self.next.init(_LangChainContextPropagationWorkflowOutboundInterceptor(outbound))
+
+    async def execute_workflow(
+        self, input: temporalio.worker.ExecuteWorkflowInput
+    ) -> Any:
+        if isinstance(input.run_fn, str):
+            name = input.run_fn
+        elif callable(input.run_fn):
+            defn = temporalio.workflow._Definition.from_run_fn(input.run_fn)
+            if defn is None or defn.name is None:
+                name = "unknown"
+            name = defn.name
+
+        with context_from_header(input, temporalio.workflow.payload_converter()):
+            # This is a sandbox friendly way to write
+            # with trace(...):
+            #   return await self.next.execute_workflow(input)
+            with temporalio.workflow.unsafe.sandbox_unrestricted():
+                t = trace(name=f"execute_workflow:{name}", run_id=workflow.info().run_id)
```

Comment:
Any concerns with tracing during replay here?

Reply:
I didn't see any duplicated traces in the UI when the replay was forced.